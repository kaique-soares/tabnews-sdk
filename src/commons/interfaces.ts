export interface TabNewsApiError {
  name: string;
  message: string;
  action: string;
  statusCode: number;
  stack?: string;
  errorId?: string;
  requestId?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  context?: any;
  errorLocationCode?: string;
  key?: string;
  type?: string;
  databaseErrorCode?: string;
}
