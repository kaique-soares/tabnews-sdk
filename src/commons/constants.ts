export const TABNEWS_BASE_URL =
  process.env.TABNEWS_BASE_URL || 'https://www.tabnews.com.br/api/v1';

export const TABNEWS_ENDPOINTS = Object.freeze({
  session: '/sessions',
  user: '/user',
  content: '/contents',
});
