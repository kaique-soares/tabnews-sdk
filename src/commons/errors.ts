import { TabNewsApiError } from './interfaces';

export class TabNewsError extends Error implements TabNewsApiError {
  action: string;
  status_code: number;
  error_id?: string;
  request_id?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  context?: any;
  error_location_code?: string;
  key?: string;
  type?: string;
  database_error_code?: string;

  constructor(tabNewsError: TabNewsApiError) {
    super(tabNewsError.message);
    this.name = tabNewsError.name;
    this.action = tabNewsError.action;
    this.status_code = tabNewsError.status_code;
    this.error_id = tabNewsError.error_id;
    this.request_id = tabNewsError.request_id;
    this.context = tabNewsError.context;
    this.stack = tabNewsError.stack;
    this.error_location_code = tabNewsError.error_location_code;
    this.key = tabNewsError.key;
    this.type = tabNewsError.type;
    this.database_error_code = tabNewsError.database_error_code;
  }
}
