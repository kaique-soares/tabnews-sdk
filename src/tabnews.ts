import { TabNewsConfig } from './interfaces';

export class TabNews {
  constructor(readonly config: TabNewsConfig = {}) {
    if (!config.credentials) {
      config.credentials = {};
    }

    config.credentials.email =
      config.credentials.email || process.env.TABNEWS_CREDENTIALS_EMAIL;
    config.credentials.password =
      config.credentials.password || process.env.TABNEWS_CREDENTIALS_PASSWORD;

    this.config = config;
  }
}
