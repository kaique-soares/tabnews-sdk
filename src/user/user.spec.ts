import 'vitest-fetch-mock';

import { expect, describe, it, beforeEach, afterEach } from 'vitest';

import { TABNEWS_ENDPOINTS, TABNEWS_HEADERS } from '@/commons';
import { TabNews } from '@/tabnews';
import {
  createTabNews,
  expectRequest,
  mockOnceApiError,
  mockOnceResponse,
  mockOnceSession,
  mockedRequest,
  resetMocks,
} from '@test/utils';

let tabNews: TabNews;

const user = {
  id: 'id',
  username: 'username',
  email: 'my@email.com',
  description: 'description',
  notifications: true,
  features: [
    'create:session',
    'read:session',
    'create:content',
    'create:content:text_root',
    'create:content:text_child',
    'update:content',
    'update:user',
  ],
  tabcoins: 9,
  tabcash: 10,
  created_at: '2022-04-05T23:50:19.920Z',
  updated_at: '2022-04-05T23:50:56.545Z',
};

describe('User', () => {
  beforeEach(() => {
    tabNews = createTabNews();
  });

  afterEach(() => {
    resetMocks();
  });

  describe('me', () => {
    it('should get me', async () => {
      mockOnceSession();

      mockOnceResponse(TABNEWS_ENDPOINTS.user, user);

      await tabNews.session.create();

      const me = await tabNews.user.me();

      expect(me).toMatchSnapshot();

      const request = mockedRequest();

      expectRequest(request).method.toBeGet();

      expectRequest(request).cookie(TABNEWS_HEADERS.sessionId).toBeDefined();
    });

    it('should throw an error when has no session', async () => {
      mockOnceApiError(TABNEWS_ENDPOINTS.user, {
        name: 'ForbiddenError',
        message: 'Usuário não pode executar esta operação.',
        action: 'Verifique se este usuário possui a feature "read:session".',
        status_code: 403,
        error_id: '1ca4236b-1a68-4573-acb8-40b220b2fd76',
        request_id: 'c17c3107-8cce-4b35-b510-0c1b5b92e3b6',
        error_location_code:
          'MODEL:AUTHORIZATION:CAN_REQUEST:FEATURE_NOT_FOUND',
      });

      expect(() => tabNews.user.me()).rejects.toThrowErrorMatchingSnapshot();

      const request = mockedRequest();

      expectRequest(request).method.toBeGet();

      expectRequest(request).cookie(TABNEWS_HEADERS.sessionId).toBeUndefined();
    });
  });
});
