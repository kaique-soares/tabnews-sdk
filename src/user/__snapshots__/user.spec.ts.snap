// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User > me > should get me 1`] = `
{
  "created_at": 2022-04-05T23:50:19.920Z,
  "description": "description",
  "email": "my@email.com",
  "features": [
    "create:session",
    "read:session",
    "create:content",
    "create:content:text_root",
    "create:content:text_child",
    "update:content",
    "update:user",
  ],
  "id": "id",
  "notifications": true,
  "tabcash": 10,
  "tabcoins": 9,
  "updated_at": 2022-04-05T23:50:56.545Z,
  "username": "username",
}
`;

exports[`User > me > should throw an error when has no session 1`] = `"Usuário não pode executar esta operação."`;
